Users/binah/access-control/access-control.cs

```diff
@@ -0,0 +1,320 @@
+
+        //if visitor is not found in list add them to the list
+        if (Visitors.TryGetValue(handle, out visitor))
+        {
+            if (DebugLogging) Log.Write("Visitor entry has been found in log.");
+        } else
+        {
+            //log user entry
+            LogEntry(agent);
+        }
+
+        if (!IsBanned(agent))
+        {
+            if(DoorsOpen || IsAdmin(agent))
+            {
+                SetAccess(agent);
+            } else
+            {
+                StartCoroutine(AskForEntry, agent);
+            }
+        } else
+        {
+            IEventSubscription timerEvent = Timer.Create(TimeSpan.FromSeconds(1), () => { Bannish(agent); });
+            Bannish(agent);
+        }
+
+    }
+
+    private void LogEntry(AgentPrivate agent)
+    {
+        Visitor visitor = new Visitor();
+        visitor.Agent = agent;
+        visitor.Name = agent.AgentInfo.Name; //used for pretty print 
+        Visitors[agent.AgentInfo.Handle.ToLower()] = visitor;
+
+        if (DebugLogging) Log.Write("Visitor entry has been logged");
+    }
+
+    private string getLogMessage()
+    {
+        string message = "There have been " + Visitors.Count + " visitors";
+        message += "\nThe doors are open: " + DoorsOpen;
+        foreach (var visitor in Visitors)
+        {
+            message += "\nName: " + visitor.Value.Name;
+            message += "\n - isAdmin: " + visitor.Value.isAdmin;
+            message += "\n - isBanned " + visitor.Value.isBanned;
+        }
+        message += "\nThe ban list: ";
+        foreach(var b in Banned)
+        {
+            message += "\n" + b;
+        }
+
+        return message;
+    }
+
+    private void ShowLog(AgentPrivate agent)
+    {
+        agent.SendChat(getLogMessage());
+    }
+
+    private void SetAccess(AgentPrivate agent)
+    {
+        bool isBanned = IsBanned(agent);
+        if (DebugLogging) Log.Write("Setting Access: " + !isBanned);
+        agent.IgnoreCollisionWith(_rb, !isBanned);
+    }
+
+    private void Bannish(AgentPrivate agent)
+    {
+        try
+        {
+            agent.Client.TeleportToUri(BannedDestination);
+            if (DebugLogging) Log.Write("Say goodbye to " + agent.AgentInfo.Name);
+        }
+        catch (NullReferenceException nre) { if (DebugLogging) Log.Write("Bannish", nre.Message); } // User Gone.
+        catch (System.Exception e) { if (DebugLogging) Log.Write("Bannish", e.ToString()); }
+    }
+
+    private void GrantEntry(AgentPrivate agent)
+    {
+        //if on ban list remove them
+        if(IsBanned(agent))
+        {
+            RemoveUserBan(agent);
+        }
+        //setAccess
+        SetAccess(agent);
+    }
+
+    private void AskForEntry(AgentPrivate agent)
+    {
+        if (DebugLogging) Log.Write("Checking Access");
+        string message = "Allow " + agent.AgentInfo.Name + " to enter?";
+        Visitor admin = Visitors[Admins[0]];
```

Comment:
there might be no admins currently in the scene

Reply:
I have a local copy unpushed that will open the doors if no admin is present, yeah I had thought about that, getting people stuck in a spawn point would be bad.